---
title: "clrDV vignette"
author: "Hongxiang Li and Tsung Fei Khang"
date: "12 April 2023"

output: rmarkdown::html_vignette
#output: pdf_document

    
vignette: >
  %\VignetteIndexEntry{introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


# Introduction

This guide provides on an overview of the R package *clrDV*, a statistical methodology for identifying genes that show differential variability (DV) between two conditions. *clrDV* is based on a compositional data analysis (CoDA) framework. The skew-normal distribution with centered parameters is used to model gene-wise null distribution of centered log-ratio (CLR) transformed RNA-Seq data. The main function for running DV test is `clrDV()`.


# Installation

Install *clrDV* from GitHub:

```{r, message=F, warning=F}
library(devtools)
install_github("Divo-Lee/clrDV")
```



# Getting Started

Load the *clrDV* package:
```{r setup}
library(clrDV)
```


# Differential Variability Analysis

We first provide an example of performing DV test on a simulated dataset of CLR-transformed RNA-Seq counts, `clrCounts2`. This dataset contains 1000 genes, with the first 100 genes exhibiting differential variability. Each group has a sample size of 200 (control vs. case). First, we load `clrCounts2`:

```{r}
data("clrCounts2") 
# 1000 genes, 200 samples per group, differential variability for the first 100 genes,
# CLR-transformed counts table
dim(clrCounts2)
clrCounts2[1:5, c(1:3, 201:203)]
```
Each row represents a gene, and each column represents a sample.

Now we can apply `clrDV()` to perform a DV test. Note that `clrDV()` does not perform the CLR-transformation itself; the CLR-transformed counts must be provided as input. Thus: 
```{r}
group2 = c(rep(0,200), rep(1,200))
clrDV_result <- clrDV(data = clrCounts2, group = group2)
head(clrDV_result, 5)
tail(clrDV_result, 5)
```
Genes with `adj_pval < 0.05` are flagged as showing statistically significant differential variability. 
```{r}
sum(clrDV_result$adj_pval < 0.05) # DV genes called
sum(clrDV_result$adj_pval[1:100] < 0.05) # true DV genes called
# observed FDR = (101-98)/101 = 0.0297; probability of Type II Error = (100-98)/100 = 0.02
```


The function `top.DV.genes()` extracts top-ranked DV genes, ranked using the SD ratio (case vs. control) of the CLR-transformed counts. Here, the `top="abs"` argument ranks the genes by $\vert\log_2($*SD_ratio*$)\vert$, that is, $\vert$*LFC*$\vert$.
```{r}
top.DV.genes(clrDV_result, top = "abs", n = 10)
```


We can use `violin.plot.clrDV()` to produce violin plots for graphically inspecting the variance of the distribution of CLR-transformed count between two groups. These plots are useful for checking that the computational results are reasonable. The violin plots in the figure below show an example of a gene that has significant DV. For gene 10, the control group has significantly  larger variance (*sigma1* = 1.415998) than the case group (*sigma2* = 0.6185685), as shown by the skew to negative values.
```{r}
violin.plot.clrDV(data = clrCounts2, "gene10", 
                  group = group2,
                  group.names = c("control", "case"))
clrDV_result[10, ]
```


The figure below shows an example of a gene that is not significant for DV. For gene 150, violin plots show that the spread of the values for both groups is about the same (*sigma1* = 1.37357, *sigma2* = 1.323281).
```{r}
violin.plot.clrDV(data = clrCounts2, "gene150", 
                  group = group2,
                  group.names = c("control", "case"))
clrDV_result[150, ]
```



The function `SN.plot()` produces a histogram of observed CLR-transformed counts, along with the fitted skew-normal probability density function for a particular gene/transcript. It can be used to graphically check how well the skew-normal distribution fits the data.
```{r}
SN.plot(clrCounts2[1, 1:200])
clr.SN.fit(clrCounts2[1, 1:200])
```


The Kolmogorov-Smirnov (KS) test can be used to compare the distribution of CLR-transformed counts of a particular gene with a skew-normal distribution. For illustration, we apply the KS test to the control group of  `clrCounts2` to evaluate whether the distribution of CLR-transformed counts in this group and the skew-normal distribution are statistically similar. The distribution of $p$-values obtained from KS-tests conducted on all 1000 genes in the control group indicates that for 995 out of 1000 genes ($99.5\%$), the skew-normal model fits the CLR-transformed count data well.

```{r,  message=F, warning=F}
library(sn) # R package for skew-normal distribution and related distributions
cp_to_dp <- function(mean=NULL, sd=NULL, skewness=NULL){
  b <- sqrt(2/pi)
  if(skewness >= 0){
    r <- (2*skewness/(4-pi))^(1/3)
  } else {
    r <- -(2*(- skewness)/(4-pi))^(1/3)
  }
  alpha <- r/sqrt(2/pi - (1-2/pi)*r^2)
  delta <- alpha/sqrt(1 + alpha^2)
  omega <- sd/sqrt(1 - (b^2)*delta^2)
  xi <- mean - b*omega*delta
  return(c(xi, omega, alpha))
} # map centered parameters to direct parameters

control_clr_SN_fit <- clr.SN.fit(clrCounts2[, 1:200]) # MLE, control group
control_sn_CP <- control_clr_SN_fit[, c("mu", "sigma", "gamma")]

control_sn_DP <- matrix(NA, nrow = dim(control_sn_CP)[1], ncol = 3)
for (i in 1:dim(control_sn_CP)[1]) {
  control_sn_DP[i,] <- c(cp_to_dp(control_sn_CP[i,1],
                                  control_sn_CP[i,2],
                                  control_sn_CP[i,3]))
}
colnames(control_sn_DP) <- c("xi", "omega", "alpha") # direct parameters
control_sn_DP <- as.data.frame(control_sn_DP)

KS_test_pvalue <- vector()
for (i in 1:dim(control_sn_CP)[1]) {
  ks <- ks.test(clrCounts2[i, 1:200],
                "psn",
                xi = control_sn_DP$xi[i],
                omega = control_sn_DP$omega[i],
                alpha = control_sn_DP$alpha[i])
  KS_test_pvalue[i] <- ks$p.value
}
sum(KS_test_pvalue < 0.05) 
# the number of genes where the skew-normal distribution fit is poor
# (1000-5)/1000 = 99.5% 
# 99.5% of the genes in the control group are well fitted by skew-normal distribution

hist(sqrt(-log10(KS_test_pvalue)), freq = F, breaks = 15,
     main = NULL, border = "grey83", 
     xlim = c(0, 1.6), ylim = c(0, 2),
     xlab = expression(sqrt(-log[10](p))))
abline(v = sqrt(-log10(0.05)), lty = 2, lwd = 1.25, col = "red")
# the dashed red line represents the p-value of 0.05
```
